### PR TITLE
llama : refactor model loading code

### PR DESCRIPTION
**Pull request by ggerganov**
- extend `gguf` API for writing
- simplify `llama.cpp` model loading / saving by using new `gguf` API
- reorder definitions in `llama.cpp` to match interface order
- merge `utils.h` in `llama.cpp`

### CODE REVIEW

**Comment by ggerganov** [comment]
@monatis  The quantization works now - I had messed up the function overloads for writing arrays / strings

**Comment by monatis** [comment]
>  The quantization works now 

I think we also need to implement writing of int arrays after adding token types yesterday

**Comment by slaren** [comment]
I am not sure if this is the right place to do this, but I think that at some point we should split `llama_context_params` into `llama_model_params` and `llama_context_params`, to make clear what parameters are used when loading a model, and what parameters can be modified per-context. For example, `use_mmap` is a model loading parameter and should be part of `llama_model_params`, but `n_ctx` should be configurable per-context in `llama_context_params`. As it is now, the users must assume that `llama_load_model_from_file` and `llama_new_context_with_model` must receive the exact same parameters in `llama_context_params`, as it is not clear which ones are relevant for each function.

File: gguf-llama.cpp [hunk 1]
```diff
@@ -1256,15 +1292,33 @@ static void llama_model_load_internal(
...
+        struct gguf_context * ctx = ml->file_loader->ctx_gguf;
+
+        hparams.n_vocab        = gguf_get_arr_n  (ctx, gguf_find_key(ctx, "tokenizer.ggml.tokens"));
+        hparams.n_ctx          = gguf_get_val_u32(ctx, gguf_find_key(ctx, "llama.context_length"));
+        hparams.n_embd         = gguf_get_val_u32(ctx, gguf_find_key(ctx, "llama.embedding_length"));
+        hparams.n_ff           = gguf_get_val_u32(ctx, gguf_find_key(ctx, "llama.feed_forward_length"));
+        hparams.n_head         = gguf_get_val_u32(ctx, gguf_find_key(ctx, "llama.attention.head_count"));
+        hparams.n_layer        = gguf_get_val_u32(ctx, gguf_find_key(ctx, "llama.block_count"));
+        hparams.n_rot          = gguf_get_val_u32(ctx, gguf_find_key(ctx, "llama.rope.dimension_count"));
+        hparams.f_rms_norm_eps = gguf_get_val_f32(ctx, gguf_find_key(ctx, "llama.rms_norm_epsilon"));
```
Comment (klosax) [inline gguf-llama.cpp:1310]: the key for rms norm is `llama.attention.layer_norm_rms_epsilon` in the spec and conversion scripts

**Comment by monatis** [comment]
I thought that we could reuse some of the code parts  for other LLMs, moving them to a common lib, but now I have the impression that things are going to a kinda self-contained structure. Is this a temporary state, or is the plan to isolate implementations of different LLMs?

**Comment by ggerganov** [comment]
My plan currently is to have `llama.cpp` implement the things that are common for all models. Ideally, when we add a new model we'll just need to implement a corresponding `struct llama_model` and `llama_build_graph()` function - these potentially can be separated in different source files, but I need to first have an example running to make a decision how to reorganize best.

The `gguf-utils.h` (formerly `llama-utils.h`) has always been "out of place". I tried merging it in `ggml`, but it does not fit really well. Maybe in the future. For now, I don't see a reason to not be inside `llama.cpp`. The `mmap` and `mlock` functionality will be shared by all models.

**Comment by ggerganov** [comment]
Haven't tested GPU and LoRA parts 

Plan is to merge this and then de-duplicate the `gguf-llama.cpp` and `llama.cpp` and make sure that all features work

**Comment by monatis** [comment]
clang++ on WSL fails to build without including `cstdarg` (for `format()`).

File: gguf-llama.cpp (hunk 1 above)
Comment (ggerganov) [reply gguf-llama.cpp:1282]: Ah, will fix it here - thanks for pointing out

**Comment by ggerganov** [comment]
> clang++ on WSL fails to build without including `cstdarg` (for `format()`).

Thanks - will fix it now.

Can you confirm that the following works as expected:

```bash
# run F16 inference
./gguf-llama-simple models/7B-v2/ggml-model-f16.gguf

# quantize
./gguf ./models/7B-v2/ggml-model-f16.gguf q
mv quant.gguf models/7B-v2/ggml-model-q4_0.gguf

# run quantized inference
./gguf-llama-simple models/7B-v2/ggml-model-q4_0.ggu
```

(replace with your respective paths

**Pull request merged**